bundles/org.openhab.core.thing/src/main/java/org/openhab/core/thing/internal/ThingHandlerCallbackImpl.java

```diff
@@ -0,0 +1,269 @@
+    }
+
+    @Override
+    public void stateUpdated(ChannelUID channelUID, State state) {
+        thingManager.communicationManager.stateUpdated(channelUID, state);
+    }
+
+    @Override
+    public void postCommand(ChannelUID channelUID, Command command) {
+        thingManager.communicationManager.postCommand(channelUID, command);
+    }
+
+    @Override
+    public void channelTriggered(Thing thing, ChannelUID channelUID, String event) {
+        thingManager.communicationManager.channelTriggered(thing, channelUID, event);
+    }
+
+    @Override
+    public void statusUpdated(Thing thing, ThingStatusInfo statusInfo) {
+        // note: all provoked operations based on a status update should be executed asynchronously!
+        ThingStatusInfo oldStatusInfo = thing.getStatusInfo();
+        ensureValidStatus(oldStatusInfo.getStatus(), statusInfo.getStatus());
+
+        if (ThingStatus.REMOVING.equals(oldStatusInfo.getStatus())
+                && !ThingStatus.REMOVED.equals(statusInfo.getStatus())) {
+            // if we go to ONLINE and are still in REMOVING, notify handler about required removal
+            if (ThingStatus.ONLINE.equals(statusInfo.getStatus())) {
+                logger.debug("Handler is initialized now and we try to remove it, because it is in REMOVING state.");
+                thingManager.notifyThingHandlerAboutRemoval(thing);
+            }
+            // only allow REMOVING -> REMOVED transition, all others are ignored because they are illegal
+            logger.debug(
+                    "Ignoring illegal status transition for thing {} from REMOVING to {}, only REMOVED would have been allowed.",
+                    thing.getUID(), statusInfo.getStatus());
+            return;
+        }
+
+        // update thing status and send event about new status
+        thingManager.setThingStatus(thing, statusInfo);
+
+        // if thing is a bridge
+        if (thing instanceof Bridge bridge) {
+            handleBridgeStatusUpdate(bridge, statusInfo, oldStatusInfo);
+        }
+        // if thing has a bridge
+        if (thing.getBridgeUID() != null) {
+            handleBridgeChildStatusUpdate(thing, oldStatusInfo);
+        }
+        // notify thing registry about thing removal
+        if (ThingStatus.REMOVED.equals(thing.getStatus())) {
+            thingManager.notifyRegistryAboutForceRemove(thing);
+        }
+    }
+
+    private void ensureValidStatus(ThingStatus oldStatus, ThingStatus newStatus) {
+        if (!(ThingStatus.UNKNOWN.equals(newStatus) || ThingStatus.ONLINE.equals(newStatus)
+                || ThingStatus.OFFLINE.equals(newStatus) || ThingStatus.REMOVED.equals(newStatus))) {
+            throw new IllegalArgumentException(
+                    MessageFormat.format("Illegal status {0}. Bindings only may set {1}, {2}, {3} or {4}.", newStatus,
+                            ThingStatus.UNKNOWN, ThingStatus.ONLINE, ThingStatus.OFFLINE, ThingStatus.REMOVED));
+        }
```

Comment:
Good change. Note to self: This will break:
- cbus
- km200
- kostalinverter

as they are setting `ThingStatus.UNINITIALIZED`. They can be fixed beforehand, only five places it seems (from GitHub search, may need to revise).